IdentityCore/src/MSIDFlightManager.m

```diff
@@ -25,33 +25,75 @@
 
 #import "MSIDFlightManager.h"
 
+@interface MSIDFlightManager()
+
+@property (nonatomic) dispatch_queue_t synchronizationQueue;
+
+@end
+
+
 @implementation MSIDFlightManager
 
 + (instancetype)sharedInstance
 {
     static MSIDFlightManager *sharedInstance = nil;
     static dispatch_once_t onceToken;
     dispatch_once(&onceToken, ^{
-        sharedInstance = [[self.class alloc] init];
+        sharedInstance = [[self.class alloc] initInternal];
     });
     
     return sharedInstance;
 }
 
+- (instancetype)initInternal
+{
+    self = [super init];
+    if (self)
+    {
+        _synchronizationQueue = [self initializeDispatchQueue];
+    }
+    return self;
+}
+
+- (dispatch_queue_t)initializeDispatchQueue
+{
+    NSString *queueName = [NSString stringWithFormat:@"com.microsoft.msidflightmanager-%@", [NSUUID UUID].UUIDString];
+    return dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_SERIAL);
+}
+
+- (void)setFlightProvider:(id<MSIDFlightManagerInterface>)flightProvider
+{
+    dispatch_barrier_async(self.synchronizationQueue, ^{
+        self->_flightProvider = flightProvider;
+    });
+}
+
 #pragma mark - MSIDFlightManagerInterface
 
 - (BOOL)boolForKey:(nonnull NSString *)flightKey 
 {
-    if (self.flightProvider) { return [self.flightProvider boolForKey:flightKey]; }
+    __block BOOL result = NO;
```

Comment:
If using concurrent queue, read can happen in parallel instead of waiting for each other. You can still keep  dispatch_barrier_async to make sure write only happened when the queue is not executing other tasks